Layer 2 Forwarding Enhancements

anilkpandey opened the pull request:
**- What I did**
1. Added code changes related to Layer 2 Forwarding Enhancements:
Added the following commands:

> VLAN range add/del
> VLAN member range add/del
> Static FDB add/del
> FDB aging time configuration
> show fdb aging time
> show mac count
> show vlan count
> added port and vlan option in 'sonic-clear fdb'

prsunny commented:
I'm not sure if these validations must be here. There is an design discussion for overall Config validator and I don't think we need to scatter these checks in multiple places.

anilkpandey commented:
Hi @stcheng and @prsunny, 

I just reverted the previous changes for validations and added a new file fdbshow for 'show mac' performance enhancements. Please review the changes.

In case you want to consider the validations changes, let me know, I will readd the changes in that case.

Thanks.

stcheng commented on config/main.py:
```diff
@@ -739,11 +739,33 @@ def add_vlan_member(ctx, vid, interface_name, untagged):
-                                                        vlan_name))
+            ctx.fail("{} is already a member of {}".format(interface_name,vlan_name))
         else:
+<<<<<<< HEAD
```
please resolve the conflicts

prsunny commented on scripts/fdbshow:
```diff
@@ -63,25 +63,51 @@ class FdbShow(object):
+            fdb_list.append(fdb)
+            pipe.hgetall(s)
+        values = pipe.execute()
+        client = self.db.redis_clients["ASIC_DB"]
```
Why is this repeated?

anilkpandey replied:
yes its not needed. will remove

prsunny commented on scripts/fdbshow:
```diff
@@ -63,25 +63,51 @@ class FdbShow(object):
+        posi = 0
+        for ent in values:
+            if 'SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID' not in ent:
+                posi = posi + 1
```
Is this case true, where SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID is not present?

anilkpandey replied:
If an entry is deleted either due to ageout or flush just after we fetched the keys, it will hit this case.

prsunny commented on scripts/fdbshow:
```diff
@@ -63,25 +63,51 @@ class FdbShow(object):
             elif 'bvid' in fdb:
-                vlan_id = port_util.get_vlan_id_from_bvid(self.db, fdb["bvid"])
-            self.bridge_mac_list.append((int(vlan_id),) + (fdb["mac"],) + (if_name,) + (fdb_type,))
+                vlan_id  = vlans[vlan_keys.index("ASIC_STATE:SAI_OBJECT_TYPE_VLAN:{}".format(fdb_list[posi][u'bvid']))] [b"SAI_VLAN_ATTR_VLAN_ID"]
```
Do you see considerable improvement over using this vs `get_vlan_id_from_bvid`? Suggest to have this logic in one place.

anilkpandey replied:
get_vlan_id_from_bvid reads from DB. But, we already have all the data retrieved from DB, so just need to parse that.

prsunny replied:
I agree to that. My point is if `get_vlan_id_from_bvid` doesn't make much of a different in performance, suggest to use the API. Any change related to BRIDGE/VLAN/BVID in future (which happened before and we had to scan all the places where `vlan `id was used) will be contained in one place rather than in multiple places.

anilkpandey replied:
understood. get_vlan_id_from_bvid is being used in 2 more places. I will see if I can change those caller also to retrieve all VLANs first and use it. Then I can modify get_vlan_id_from_bvid to not read from DB. Thanks.

prsunny replied:
How much performance advantage are we getting without the call? Do you have any numbers?

anilkpandey replied:
For 32K entries, it takes 7 seconds if I don't call get_vlan_id_from_bvid and it takes 5 min if I call get_vlan_id_from_bvid.

prsunny replied:
oh ok, that's quite an improvement. Lets keep the change then!

prsunny commented:
retest this please

jleveque commented:
Retest this please

jleveque commented:
Retest this please

anilkpandey commented:
Added code changes related to Layer 2 Forwarding Enhancements

jleveque commented:
Retest this please

anilkpandey commented:
@prsunny @stcheng , can you please complete this review? Regarding the test failures, the vnet, vrf and subport tests are failing for all PRs in this repo. Any ideas whats the reason?

anilkpandey commented:
Retest this please

anilkpandey commented:
retest this please

anilkpandey commented:
retest this please

ben-gale commented:
@prsunny - can this one be merged no pls?

prsunny commented on scripts/fast-reboot-dump.py:
```diff
@@ -147,6 +147,108 @@ def get_fdb(db, vlan_name, vlan_id, bridge_id_2_iface):
 
     return fdb_entries, available_macs, map_mac_ip
 
+def generate_fdb_entries_2(filename):
```
Why do we have this change in fast-reboot?. @tahmed-dev for viz. fast-reboot enhancement is not mentioned in description. May require additional testing with some recent changes in this area

tahmed-dev replied:
is this a replacement of the generate_dfb_entries? if yes, what is the time saving here when using this version?

jainp1979 replied:
Here, written and used optimized version (generate_fdb_entries_2) of generate_fdb_entries 
1) Modifed the functionality to work based on available mac addresses rather than available vlans 
2) used bulk (pipe) asic-db apis to get vlan and fdb from ASIC db instead of getting the data for individual vlan/fdb entry 
3) stored vlan_oid to vlan_id and vice-versa mapping in local memory instead of getting the same everytime from ASIC-DB 
4) done away with get_fdb function to be able to modify the overall logic and reduce copying of objects and function calls.

measured specifically fast-reboot-dump.py script from FDB perspective and now it takes around 10 sec instead of 12 minutes with 4K empty vlans;
Also tested the modified script with 4k mac addresses in 4k vlans (1 mac per vlan) and it gets completed with in 10 sec.

prsunny commented on config/main.py:
```diff
@@ -1828,6 +1841,113 @@ def delete_snmptrap_server(ctx, ver):
+    db = ctx.obj['db']
+
+    mac_valid = bool(re.match('^' + '[\:\-]'.join(['([0-9a-f]{2})']*6) + '$', mac.lower()))
+    if mac_valid == False:
```
can we have mac validation in a separate function which can help future use?

prsunny commented on config/main.py:
```diff
@@ -1828,6 +1841,113 @@ def delete_snmptrap_server(ctx, ver):
+    if interface_name_is_valid(interface_name) is False:
+        ctx.fail("Interface name is invalid!!")
+
+    db.set_entry('FDB', (vlan_name, mac), {'port': interface_name })
```
Currently, there is no config db support for FDB and aging time. For this, we need to to have this [PR](https://github.com/Azure/sonic-swss/pull/885), correct?

prsunny commented on show/main.py:
```diff
@@ -716,22 +716,59 @@ def pwm_buffer_pool():
-    cmd = "fdbshow"
+    # Fetching data from config_db for SWITCH
+    switch_table = config_db.get_table('SWITCH')
+    switch_keys = switch_table.keys()
```
can we have this displayed from APP db? So even if there is config, it will display the default.

tahmed-dev commented on config/vlan.py:
```diff
@@ -201,3 +201,333 @@ def del_vlan_dhcp_relay_destination(db, vid, dhcp_relay_destination_ip):
+    warning_ip_list = []
+    clients = db.redis_clients["CONFIG_DB"]
+    pipe = clients.pipeline()
+    vlan_member_keys = db.keys('CONFIG_DB', "*VLAN_MEMBER*")
```
suggest to use `clients.scan(...)` API

tahmed-dev replied:
@anilkpandey [client scan](https://redis.io/commands/scan) it use cursor iterator so it does not lock the redis db for long time. [Here](https://github.com/Azure/sonic-py-swsssdk/blob/master/src/swsssdk/configdb.py#L429) is a code that uses client scan in swsssdk.

tahmed-dev commented on scripts/fast-reboot-dump.py:
```diff
@@ -147,6 +147,108 @@ def get_fdb(db, vlan_name, vlan_id, bridge_id_2_iface):
+
+    #print("generate_fdb_entries_2 before fdb-key-getall" + datetime.datetime.now().strftime("%H:%M:%S.%f"))
+    pipe = client.pipeline()
+    fdb_list = db.keys(db.ASIC_DB, "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:*")
```
suggest to use `client.scan(...)` API

anilkpandey commented:
retest this please

ben-gale commented:
> overall, lgtm, seems unit test is missing, requesting few more reviewers

@prsunny - the unit tests are covered in https://github.com/Azure/sonic-swss/pull/885 - however these can't pass without this PR (#529) merging. It also seems like Tamer has reviewed, so this covers your request. Can this one be merged now please?

lguohan commented:
@ben-gale, sonic utilities recently added unit test infrastructure. now all new features needs unit test. you can checkout the coverage report here. https://sonic-jenkins.westus2.cloudapp.azure.com/job/common/job/sonic-utilities-build/RCov_20Report/

it is important that we are adding unit test for new features to avoid regression.

anilkpan commented:
retest this please

anilkpan commented:
retest this please

lguohan commented on scripts/fdbshow:
```diff
@@ -31,8 +31,8 @@ import json
 from natsort import natsorted
-from swsssdk import port_util
-from swsscommon.swsscommon import SonicV2Connector
+from swsssdk import SonicV2Connector, port_util
```
we should not use swsssdk, it is deprecating.

anilkpandey commented:
/azp run

anilkpandey commented:
retest this please

anilkpandey commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

anilkpandey commented:
/azpw run

mssonicbld commented:
/AzurePipelines run
